sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,232 @@
+# Author: Henry Lin <hlin117@gmail.com>
+
+from __future__ import division, absolute_import
+
+import numbers
+import numpy as np
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils.validation import check_array, check_is_fitted, column_or_1d
+
+
+class KBinsDiscretizer(BaseEstimator, TransformerMixin):
+    """Bins continuous data into k equal width intervals.
+
+    Parameters
+    ----------
+    n_bins : int or array-like, shape (n_features_,) (default=2)
+        The number of bins to produce. The intervals for the bins are
+        determined by the minimum and maximum of the input data.
+        Raises ValueError if n_bins < 2.
+
+        If n_bins is an array, and there is a categorical feature at index i,
+        n_bins[i] will be ignored.
+
+    categorical_features : int array-like (default=None)
+        Column indices of categorical features. Categorical features
+        will not be discretized and will be ignored by the transformation.
+
+    Attributes
+    ----------
+    categorical_features_ : int array, shape (n_features_,)
+        Indices of categorical features. These are not affected by the
+        transformation.
+
+    min_ : float array, shape (n_features_,)
+        Minimum value per feature in X. A categorical feature at index i will
+        have min_[i] == 0.
+
+    ptp_ : float array, shape (n_features_,)
+        X.max(axis=0) - X.min(axis=0). A categorical feature at index i will
+        have ptp_[i] == 1.
+
+    clip_min_ : float array, shape (n_features_,)
```

Comment:
The argument for calculating `clip_min` and `clip_max` in `fit` is because these attributes can be shared across calls to `transform`. On the other hand, these attributes aren't very useful for the user.

Let's change `clip_min` and `clip_max` to be private variables.